### PR TITLE
Fix domain regexp

### LINKED ISSUE #37: issues with allowed domain regexp
When spider downloads the first page it updates the regexp for allowed domains. Currently the code has some issues:

* If the domain strats with `www` then non-www version gets banned. I think we should allow non-www version in this case (to get pages which are at non-www links), but handle it at duplicate filter level (to avoid getting the same content twice if a website has www and non-www mirrors).
* URL which is passed to re.compile is not escaped; it means `.` in domain name matches any character, not only a dot.

### PR DESCRIPTION
**Pull request by lopuhin**
This should fix #37 - it correctly normalizes and escapes initial allowed regexp, and adds a duplicate filter that does not take www and http/https into account.
Also, I'm not super sure here, but I maybe the duplicate filter from scrapy-splash should apply canonicalize_url without removing fragments to ``request.meta['splash']['args']['url']`` when calculating request fingerprint.

### CODE REVIEW

**Comment by kmike** [comment]
@lopuhin yeah, a nice catch about canonicalize_url in scrapy-splash. It should take fragment in account.

File: undercrawler/spiders/base_spider.py [hunk 1]
```diff
@@ -362,3 +357,25 @@ def link_to_url(link):
     if link.fragment and link.fragment != '#':
         return "#".join([link.url, link.fragment])
     return link.url
+
+
+def allowed_re(url, hard_url_constraint):
```
Comment (kmike) [inline undercrawler/spiders/base_spider.py]: What do you think about allowing all subdomains by default, not only www?

File: undercrawler/dupe_filter.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+from scrapy_splash import SplashAwareDupeFilter
+
+
+class DupeFilter(SplashAwareDupeFilter):
+    """
+    Consider same urls with and without www as duplicates.
+    """
+    def request_fingerprint(self, request):
+        if not request.meta.get('_splash_processed'):
+            # canonicalize_url required only due to a bug (?) in scrapy_splash
```
Comment (kmike) [inline undercrawler/dupe_filter.py]: This should be fixed in https://github.com/scrapy-plugins/scrapy-splash/commit/8ca86351b0a6a40b05b58f096e09650de142beed.

Comment (lopuhin) [inline undercrawler/dupe_filter.py]: Thanks! I removed the ``canonicalize_url`` here.

File: undercrawler/spiders/base_spider.py (hunk 1 above)
Comment (lopuhin) [inline undercrawler/spiders/base_spider.py]: Yes, I think it makes sense to allow all subdomains lower then the start subdomain, I'll update the regexp.

Comment (lopuhin) [inline undercrawler/spiders/base_spider.py]: @kmike I've updated the regexp to allow subdomains: it's a little bit more permissive than what is allowed by domain spec, but I believe there are not valid urls that could be handled incorrectly.

**Pull request merged**